### DISCUSSION TITLE
Unable to make it work on Raspberry Pi

### DISCUSSION THREAD

**Discussion by lorygoth** [dscripka/openWakeWord#23] [Q&A] [answered]
I have raspberry pi 4, running 64bit Raspberry OS and a USB mic, which was working fine with Rhasspy docker container, and I've checked also mic input from sounddevice.

I wrote a simple python script to try it locally: https://gist.github.com/lorygoth/f669bc2318521bd2dc4ceb8f12440055
But as I run it I don't get any correct prediction levels in a reaction to any input. Could anyone please have a look at the script?
Also I have other warnings in the very beginning of script running, maybe it's also important:

```
2023-04-10 14:30:47.961337967 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '74'. It is not used by any node and should be removed from the model.
2023-04-10 14:30:47.961464022 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '71'. It is not used by any node and should be removed from the model.
2023-04-10 14:30:47.961508688 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '61'. It is not used by any node and should be removed from the model.
2023-04-10 14:30:47.961550392 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '58'. It is not used by any node and should be removed from the model.
/home/pi/.local/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py:54: UserWarning: Specified provider 'CUDAExecutionProvider' is not in available provider names.Available providers: 'CPUExecutionProvider'
  warnings.warn(
found Yamaha YVC-200 device id: 3. sample rate: 48000.0
^CTraceback (most recent call last):
  File "/home/pi/openwakeword/test.py", line 87, in <module>
    prediction = oww.predict(data)
  File "/home/pi/.local/lib/python3.9/site-packages/openwakeword/model.py", line 190, in predict
    self.preprocessor(x)
  File "/home/pi/.local/lib/python3.9/site-packages/openwakeword/utils.py", line 311, in __call__
    self._streaming_features(x)
  File "/home/pi/.local/lib/python3.9/site-packages/openwakeword/utils.py", line 298, in _streaming_features
    self._streaming_melspectrogram(x)
  File "/home/pi/.local/lib/python3.9/site-packages/openwakeword/utils.py", line 288, in _streaming_melspectrogram
    (self.melspectrogram_buffer, self._get_melspectrogram(list(self.raw_data_buffer)[-len(x)-160*3:]))
  File "/home/pi/.local/lib/python3.9/site-packages/openwakeword/utils.py", line 95, in _get_melspectrogram
    outputs = self.melspec_model.run(None, {'input': x})
  File "/home/pi/.local/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 200, in run
    return self._sess.run(output_names, input_feed, run_options)
KeyboardInterrupt
```

**Comment by dscripka** [comment, marked as answer]
Hey @lorygoth, thanks for making an example based on sounddevice. I can think of cases where that would be preferred over PyAudio.

It looks like the issue is with data types. On line #32 of your example script, you are getting the data from the callback function, but since you are using a RawInputStream from sounddevice, this will return a Python buffer object. Using `np.asarray` will not properly convert the buffer object, you will need to use `np.frombuffer(indata, dtype=np.int16)`. Also, on line #41 you use `np.asarray` to convert a list to a Numpy array. This works, but will use the default data type for numpy arrays, `float64`. So you'll need to convert that back into `np.int16` which is what openWakeWord requires.

After I make these changes I get the expected behavior with your script and the detection works properly.

**Reply by lorygoth** [reply]
Works like a charm, thank you so much!
Updated the gist so others could also have an example.

**Reply by dalehumby** [reply]
Hi @lorygoth, I also saw your [Rhasspy forum](https://community.rhasspy.org/t/rhasspy-3-developer-preview/4409/57) post and wondered if you managed to get OpenWakeWord to trigger Rhasspy? (Either 2.5 or 3?) And if so, how did you achieve that? 

I've been thinking of [Dockerising OWW](https://github.com/dscripka/openWakeWord/discussions/25) to make it simpler to use with Rhasspy 2.5 and 3.

**Reply by lorygoth** [reply]
Actually yes, I have now a small python app, which runs in docker and listens for the mic input instead of rhasspy. To trigger rhasspy I call wake word detection over Hermes mqtt. This app also sends chunks of audio.

I didn't yet put it to github, because I also want to add some voice quality improvement logic, and I'm still on it.

But in the end it should be a very easy configurable docker image, which will listen to the wake word and provide audio for voice recognition. And as a bonus - audio will be sent only when voice recognition is ongoing.

**Reply by dalehumby** [reply]
Sounds great! I've also had a go at Dockerising OWW (https://github.com/dalehumby/openWakeWord-rhasspy) but I would really like to see what you have when you get it up on GitHub. Please @ me when you do.

I chose to use audio recording / the mic in Rhasspy, stream audio via UDP to the Docker container which does the detection and triggers Rhasspy via Hermes/MQTT. I did this because I would like to run 1 central instance of the OWW service on my base for all of the satellites to use, instead of running OWW on each satellite.

With a central service, if multiple satellites trigger simultaneously then I should be able to arbitrate and only trigger the device closest to the speaker.

**Reply by dscripka** [reply]
I'd also be interested in seeing what you've developed @lorygoth, glad that you've found a way to get it working with Rhasspy.

Docker seems like a good solution in general, but for Rhasspy 3 in particular it looks like it would be quite simple to make an appropriate wrapper (using [this example with Pocupine](https://github.com/rhasspy/rhasspy3/blob/master/programs/wake/porcupine1/bin/porcupine_raw_text.py) as a guide).
